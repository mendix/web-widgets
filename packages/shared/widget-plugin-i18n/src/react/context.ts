import { createContext } from "react";
import { PluginI18n } from "../types";

const contextPath = "com.mendix.widget.plugin.i18n.context" as const;

type I18nContext = React.Context<PluginI18n | null>;

const createI18nContext = (): I18nContext => createContext<PluginI18n | null>(null);

declare global {
    interface Window {
        [contextPath]?: ReturnType<typeof createI18nContext>;
    }
}

export function getGlobalI18nPluginContext(): I18nContext {
    return window[contextPath] ?? createI18nContext();
}
