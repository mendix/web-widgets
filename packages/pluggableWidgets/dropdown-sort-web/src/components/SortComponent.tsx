import { createElement, CSSProperties, ReactElement, useCallback, useEffect, useRef, useState } from "react";
import { useOnClickOutside } from "@mendix/widget-plugin-hooks/useOnClickOutside";
import { usePositionObserver } from "@mendix/widget-plugin-hooks/usePositionObserver";
import { SortDirection } from "@mendix/widget-plugin-sorting";
import classNames from "classnames";
import { createPortal } from "react-dom";

export interface SortOption {
    caption: string;
    value: string;
}

interface SortComponentProps {
    className?: string;
    defaultDirection?: SortDirection;
    defaultOption?: SortOption;
    emptyOptionCaption?: string;
    id?: string;
    options: SortOption[];
    tabIndex?: number;
    screenReaderButtonCaption?: string;
    screenReaderInputCaption?: string;
    styles?: CSSProperties;
    updateSort?: (value: SortOption, direction: SortDirection) => void;
}

export function SortComponent(props: SortComponentProps): ReactElement {
    const [valueInput, setValueInput] = useState(props.defaultOption?.caption ?? props.emptyOptionCaption ?? "");
    const [selectedSort, setSelectedSort] = useState<SortOption>(
        props.defaultOption ?? {
            caption: props.emptyOptionCaption ?? "",
            value: ""
        }
    );
    const [show, setShow] = useState(false);
    const [dropdownWidth, setDropdownWidth] = useState(0);
    const [direction, setDirection] = useState(props.defaultDirection ?? "asc");

    const componentRef = useRef<HTMLDivElement>(null);
    const optionsRef = useRef<HTMLUListElement>(null);

    const position = usePositionObserver(componentRef.current, show);

    const onClick = useCallback((option: SortOption) => {
        setValueInput(option.caption);
        setSelectedSort(option);
        setShow(false);
    }, []);

    useOnClickOutside([componentRef, optionsRef], () => setShow(false));

    useEffect(() => {
        if (selectedSort) {
            props.updateSort?.(selectedSort, direction);
        }
    }, [direction, selectedSort]);

    const showPlaceholder = !selectedSort || valueInput === props.emptyOptionCaption;

    const optionsComponent = createPortal(
        <ul
            ref={optionsRef}
            id={`${props.id}-dropdown-list`}
            className="dropdown-list"
            role="menu"
            data-focusindex={0}
            style={{ position: "fixed", width: dropdownWidth, top: position?.bottom, left: position?.left }}
        >
            {props.options.map((option, index) => (
                <li
                    className={classNames({
                        "filter-selected": selectedSort?.value === option.value
                    })}
                    key={index}
                    onClick={e => {
                        e.preventDefault();
                        e.stopPropagation();
                        onClick(option);
                    }}
                    onKeyDown={e => {
                        if (e.key === "Enter" || e.key === " ") {
                            e.preventDefault();
                            e.stopPropagation();
                            onClick(option);
                        } else if (e.key === "Tab" && index + 1 === props.options.length) {
                            e.preventDefault();
                            setShow(false);
                            componentRef.current?.querySelector("button")?.focus();
                        } else if ((e.key === "Tab" && e.shiftKey && index === 0) || e.key === "Escape") {
                            e.preventDefault();
                            setShow(false);
                            componentRef.current?.querySelector("input")?.focus();
                        }
                    }}
                    role="menuitem"
                    tabIndex={0}
                >
                    <div className="filter-label">{option.caption}</div>
                </li>
            ))}
        </ul>,
        document.body
    );

    const containerClick = useCallback(() => {
        setShow(show => !show);
        setTimeout(() => {
            (optionsRef.current?.querySelector("li.filter-selected") as HTMLElement)?.focus();
        }, 10);
    }, []);

    return (
        <div
            className={classNames("dropdown-container", props.className)}
            data-focusindex={props.tabIndex ?? 0}
            ref={componentRef}
            style={props.styles}
        >
            <div className="dropdown-triggerer-wrapper">
                <input
                    value={showPlaceholder ? "" : valueInput}
                    placeholder={showPlaceholder ? props.emptyOptionCaption : undefined}
                    className="form-control dropdown-triggerer"
                    onClick={containerClick}
                    onKeyDown={e => {
                        if (e.key === "Enter" || e.key === " ") {
                            e.preventDefault();
                            e.stopPropagation();
                            containerClick();
                        }
                    }}
                    aria-haspopup
                    ref={inputRef => {
                        if (inputRef && inputRef.clientWidth) {
                            setDropdownWidth(inputRef.clientWidth);
                        }
                    }}
                    aria-expanded={show}
                    aria-controls={`${props.id}-dropdown-list`}
                    aria-label={props.screenReaderInputCaption}
                />
                <button
                    aria-label={props.screenReaderButtonCaption}
                    className={classNames("btn btn-default btn-sort", {
                        "icon-asc": direction === "asc",
                        "icon-desc": direction === "desc"
                    })}
                    onClick={() => setDirection(prev => (prev === "asc" ? "desc" : "asc"))}
                />
            </div>
            {show && optionsComponent}
        </div>
    );
}
