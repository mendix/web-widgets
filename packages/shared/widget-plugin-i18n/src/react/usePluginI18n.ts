import { useState } from "react";
import { I18nController } from "../controllers/I18nController";
import { PluginI18n } from "../types";

export function usePluginI18n(languageCode: string): PluginI18n | null {
    const [plugin, setPlugin] = useState<PluginI18n | null>(() => {
        new I18nController({
            languageCode,
            onPluginChange: plugin => setPlugin(plugin)
        });

        return null;
    });

    return plugin;
}
