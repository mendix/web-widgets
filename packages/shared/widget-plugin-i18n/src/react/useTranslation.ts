import { useContext } from "react";
import { PluginI18n } from "../types";
import { getGlobalI18nPluginContext } from "./context";

export function useTranslation(): PluginI18n | null {
    return useContext(getGlobalI18nPluginContext());
}
