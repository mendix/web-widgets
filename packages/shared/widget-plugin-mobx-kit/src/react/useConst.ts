import { useRef } from "react";

export function useConst<T>(fn: T): T extends (...args: never) => infer R ? R : T {
    return (useRef<T | null>(null).current ??= typeof fn === "function" ? fn() : fn);
}
