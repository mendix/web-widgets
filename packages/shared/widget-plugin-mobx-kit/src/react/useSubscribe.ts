import { IReactionOptions, reaction } from "mobx";
import { useEffect, useState } from "react";

export function useSubscribe<Store extends object, T, FireImmediately extends boolean = false>(
    init: Store | (() => Store),
    selector: (store: Store) => T,
    options?: IReactionOptions<T, FireImmediately>
): [T, Store] {
    const [store] = useState(() => (typeof init === "function" ? init() : init));
    const [value, setValue] = useState(() => selector(store));

    function setup(): () => void {
        return reaction(() => selector(store), setValue, options);
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
    useEffect(setup, []);

    return [value, store];
}
