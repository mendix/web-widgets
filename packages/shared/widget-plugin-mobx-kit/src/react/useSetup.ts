import { useEffect } from "react";
import { useConst } from "./useConst";

export function useSetup<
    T extends {
        setup(): void | (() => void);
    }
>(fn: () => T): T {
    const obj = useConst(fn);
    useEffect(() => obj.setup(), [obj]);
    return obj;
}
