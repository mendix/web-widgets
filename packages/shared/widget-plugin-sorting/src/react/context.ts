import { useConst } from "@mendix/widget-plugin-mobx-kit/react/useConst";
import { generateUUID } from "@mendix/widget-plugin-platform/framework/generate-uuid";
import { Context, createContext, useContext, useEffect } from "react";
import { SortStoreHost } from "../stores/SortStoreHost";
import { Result, error, value } from "./result-meta";

export interface SortAPI {
    version: 1;
    host: SortStoreHost;
}

const SORT_PATH = "com.mendix.widgets.web.sortable.sortContext";

export function getGlobalSortContext(): Context<SortAPI | null> {
    const scope = window.top === window ? window : window.top;
    return ((scope as any)[SORT_PATH] ??= createContext<SortAPI | null>(null));
}

export function useSortAPI(): Result<SortAPI, Error> {
    const api = useContext(getGlobalSortContext());
    if (api === null) {
        return error(new Error("Error: widget is out of context. Please place the widget inside the Gallery header."));
    }
    return value(api);
}

export function useLockSortAPI(api: SortAPI): Result<SortAPI, Error> {
    const id = useLock(api);

    if (api.host.usedBy !== id) {
        return error(
            new Error(
                `Error: Sort API is already in use by another widget. Remove other sort widgets and refresh the page.`
            )
        );
    }

    return value(api);
}

function useLock({ host }: SortAPI): string {
    const [unlock, id] = useConst(() => {
        const id = `useLock@${generateUUID()}`;
        return [host.lock(id), id] as const;
    });

    useEffect(() => unlock, [unlock]);

    return id;
}
