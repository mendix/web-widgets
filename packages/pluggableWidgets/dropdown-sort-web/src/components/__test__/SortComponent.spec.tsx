import "@testing-library/jest-dom";
import { act, fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { SortComponent } from "../SortComponent";

const defaultOptions: Array<{ caption: string; value: string }> = [
    { caption: "Empty option", value: "none" },
    { caption: "1", value: "_1" },
    { caption: "2", value: "_2" },
    { caption: "3", value: "_3" }
];

jest.useFakeTimers();

describe("Sort selector", () => {
    describe("renders correctly", () => {
        it("with options", () => {
            const component = render(<SortComponent options={defaultOptions} direction="asc" value={null} />);

            expect(component.container).toMatchSnapshot();
        });

        it("with no options", () => {
            const component = render(<SortComponent options={[]} direction="asc" value={null} />);

            expect(component.container).toMatchSnapshot();
        });

        it("with a11y properties", () => {
            const component = render(
                <SortComponent
                    options={defaultOptions}
                    direction="asc"
                    value={null}
                    screenReaderButtonCaption="my button"
                    screenReaderInputCaption="my input"
                />
            );

            expect(component.container).toMatchSnapshot();
        });

        it("with placeholder", () => {
            const component = render(
                <SortComponent options={defaultOptions} placeholder={"find me"} direction="asc" value={null} />
            );

            expect(component.container).toMatchSnapshot();
            expect(component.asFragment().querySelector("input")?.getAttribute("placeholder")).toBe("find me");
        });
    });

    it("selects default option", () => {
        const defaultOption = defaultOptions[1];
        const { asFragment } = render(
            <SortComponent options={defaultOptions} value={defaultOption.value} direction="asc" />
        );
        expect(asFragment().querySelector("input")?.getAttribute("value")).toBe(defaultOption.caption);
    });

    describe("when value changes", () => {
        it("calls onSelect when value changes", async () => {
            const onSelectHandler = jest.fn();
            const component = render(
                <SortComponent
                    placeholder="TestInput"
                    options={defaultOptions}
                    onSelect={onSelectHandler}
                    direction="asc"
                    value={null}
                />
            );

            const input = component.container.querySelector("input");
            await fireEvent.click(input as Element);

            const items = component.queryAllByRole("menuitem");
            await fireEvent.click(items[0]);

            expect(onSelectHandler).toHaveBeenCalled();
        });

        it("shows selected option on input value", async () => {
            let defaultOption = defaultOptions[1];
            const component = render(
                <SortComponent
                    options={defaultOptions}
                    value={defaultOption.value}
                    direction="asc"
                    onSelect={optionValue => {
                        defaultOption = defaultOptions.find(opt => opt.value === optionValue) || defaultOption;
                    }}
                />
            );

            const input = component.container.querySelector("input");
            fireEvent.click(input as Element);

            expect(input?.getAttribute("value")).toBe(defaultOption.caption);

            const item = component.queryAllByRole("menuitem")[defaultOptions.length - 1];
            fireEvent.click(item);

            component.rerender(
                <SortComponent
                    options={defaultOptions}
                    value={defaultOption.value}
                    direction="asc"
                    onSelect={optionValue => {
                        defaultOption = defaultOptions.find(opt => opt.value === optionValue) || defaultOption;
                    }}
                />
            );
            expect(component.container.querySelector("input")?.getAttribute("value")).toBe(defaultOptions[3].caption);
        });
    });

    describe("focus", () => {
        it("changes focused element when pressing the input", async () => {
            render(<SortComponent options={defaultOptions} placeholder="Click me" direction="asc" value={null} />);
            expect(document.body).toHaveFocus();
            const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });

            const input = screen.getByPlaceholderText("Click me");
            expect(input).toBeDefined();

            await user.click(input);

            // Wait for setTimeout(10ms) to execute the focus logic
            act(() => {
                jest.advanceTimersByTime(20);
            });

            const items = screen.queryAllByRole("menuitem");
            expect(items[0]).toHaveFocus();
        });

        it("changes focused element back to the input when pressing shift+tab in the first element", async () => {
            render(<SortComponent options={defaultOptions} placeholder="Click me" direction="asc" value={null} />);
            expect(document.body).toHaveFocus();
            const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });

            const input = screen.getByPlaceholderText("Click me");
            expect(input).toBeDefined();
            await user.click(input);

            // Wait for setTimeout(10ms) to execute the focus logic
            act(() => {
                jest.advanceTimersByTime(20);
            });

            const items = screen.getAllByRole("menuitem");
            expect(items[0]).toHaveFocus();

            await user.tab({ shift: true });

            // Wait for focus change
            act(() => {
                jest.advanceTimersByTime(20);
            });

            expect(input).toHaveFocus();
        });

        it("changes focused element back to the input when pressing tab on the last item", async () => {
            render(
                <SortComponent
                    options={[
                        { caption: "Click me", value: "click_me" },
                        { caption: "1", value: "_1" }
                    ]}
                    placeholder="Click me"
                    direction="asc"
                    value={null}
                />
            );

            expect(document.body).toHaveFocus();
            const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });

            const input = screen.getByPlaceholderText("Click me");
            await user.click(input);

            // Wait for setTimeout(10ms) to execute the focus logic
            act(() => {
                jest.advanceTimersByTime(20);
            });

            const items = screen.getAllByRole("menuitem");
            expect(items[0]).toHaveFocus();

            await user.tab();
            expect(items[1]).toHaveFocus();
            await user.tab();

            // Wait for focus change
            act(() => {
                jest.advanceTimersByTime(20);
            });

            const button = screen.getByRole("button");

            expect(button).toHaveFocus();
        });

        it("changes focused element back to the input when pressing escape on any item", async () => {
            render(
                <SortComponent
                    options={[
                        { caption: "Click me", value: "click_me" },
                        { caption: "1", value: "_1" },
                        { caption: "2", value: "_2" }
                    ]}
                    placeholder="Click me"
                    direction="asc"
                    value={null}
                />
            );
            expect(document.body).toHaveFocus();
            const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });

            const input = screen.getByPlaceholderText("Click me");
            await user.click(input);

            // Wait for setTimeout(10ms) to execute the focus logic
            act(() => {
                jest.advanceTimersByTime(20);
            });

            const items = screen.getAllByRole("menuitem");
            expect(items).toHaveLength(3);
            expect(items[0]).toHaveFocus();

            await user.tab();

            expect(items[1]).toHaveFocus();

            await user.keyboard("{Escape}");

            // Wait for focus change
            act(() => {
                jest.advanceTimersByTime(20);
            });

            expect(input).toHaveFocus();
        });
    });
});
