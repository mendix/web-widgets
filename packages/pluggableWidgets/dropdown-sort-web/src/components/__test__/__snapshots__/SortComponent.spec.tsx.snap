// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort selector renders correctly with a11y properties 1`] = `
<div>
  <div
    class="dropdown-container"
    data-focusindex="0"
  >
    <div
      class="dropdown-triggerer-wrapper"
    >
      <input
        aria-controls="undefined-dropdown-list"
        aria-expanded="false"
        aria-haspopup="true"
        aria-label="my input"
        class="form-control dropdown-triggerer"
        value=""
      />
      <button
        aria-label="my button"
        class="btn btn-default btn-sort icon-asc"
      />
    </div>
  </div>
</div>
`;

exports[`Sort selector renders correctly with no options 1`] = `
<div>
  <div
    class="dropdown-container"
    data-focusindex="0"
  >
    <div
      class="dropdown-triggerer-wrapper"
    >
      <input
        aria-controls="undefined-dropdown-list"
        aria-expanded="false"
        aria-haspopup="true"
        class="form-control dropdown-triggerer"
        value=""
      />
      <button
        class="btn btn-default btn-sort icon-asc"
      />
    </div>
  </div>
</div>
`;

exports[`Sort selector renders correctly with options 1`] = `
<div>
  <div
    class="dropdown-container"
    data-focusindex="0"
  >
    <div
      class="dropdown-triggerer-wrapper"
    >
      <input
        aria-controls="undefined-dropdown-list"
        aria-expanded="false"
        aria-haspopup="true"
        class="form-control dropdown-triggerer"
        value=""
      />
      <button
        class="btn btn-default btn-sort icon-asc"
      />
    </div>
  </div>
</div>
`;

exports[`Sort selector renders correctly with placeholder 1`] = `
<div>
  <div
    class="dropdown-container"
    data-focusindex="0"
  >
    <div
      class="dropdown-triggerer-wrapper"
    >
      <input
        aria-controls="undefined-dropdown-list"
        aria-expanded="false"
        aria-haspopup="true"
        class="form-control dropdown-triggerer"
        placeholder="find me"
        value=""
      />
      <button
        class="btn btn-default btn-sort icon-asc"
      />
    </div>
  </div>
</div>
`;
